pkg/apis/certmanager/v1alpha1/types_order.go

```diff
@@ -192,6 +192,19 @@ type HTTP01SolverConfig struct {
 	// without an ingress class set will be created to solve HTTP01 challenges.
 	// If this field is specified, 'ingress' **must not** be specified.
 	IngressClass *string `json:"ingressClass,omitempty"`
+
+	// InternalSelfCheck flag allows to make acme challenge self-check successful
+	// when it is not possible to test reachability from internet. For example,
+	// 1) you have a service exposed with internal ingress controller (not reachable from internet),
+	// 2) you want to secure it by serving acme challenge with external ingress controller,
+	// 3) external controller has internal IP (.Status.LoadBalancer.Ingress[0].IP),
+	// 4) you have external IP translated (NAT) to internal IP,
+	// 5) external IP is not reachable from cluster,
+	// - in this case with InternalSelfCheck enabled self-check request will be sent
+	// to internal IP of external ingress controller with Host header set. So self-check
+	// will ensure that acme solver service is ready but external controller reachability
+	// from internet and DNS configuration will not be checked.
+	InternalSelfCheck bool `json:"internalSelfCheck,omitempty"`
```

Comment:
I think we're better off exposing this as some kind of enum-like field, something like:

```
selfCheckHostSource:
  ingressStatus:
    type: IPAddress
```

Not 100% certain on the API surface for it yet though, it'd be great if we could put together a short proposal (just a couple of paragraphs) on the semantics, so they can be discussed 😄